server/core/dependency_injector.py

```diff
@@ -0,0 +1,137 @@
+    assert classes["other"].some_class is classes["some_class"]
+    ```
+
+    """
+
+    def __init__(self) -> None:
+        # Objects which are available to the constructors of injected objects
+        self.injectables: Dict[str, object] = {}
+
+    def add_injectables(self, **kwargs: object) -> None:
+        """
+        Register additional objects that can be requested by injected classes.
+        """
+        self.injectables.update(kwargs)
+
+    def build_classes(self, classes: Dict[str, type]) -> Dict[str, object]:
+        """
+        Resolve dependencies by name and instantiate each class.
+        """
+        dep = self._make_dependency_graph(classes)
+        # Can get away with a shallow copy because dep values are not modified
+        # in-place.
+        param_map = dep.copy()
+
+        return self._build_classes_from_dependencies(dep, classes, param_map)
+
+    def _make_dependency_graph(self, classes: Dict[str, type]) -> DependencyGraph:
+        """Build dependency graph"""
+        graph: DependencyGraph = defaultdict(list)
+        for name in self.injectables:
+            graph[name] = []
+
+        for obj_name, klass in classes.items():
+            signature = inspect.signature(klass.__init__)
+            # Strip off the `self` parameter
+            params = list(signature.parameters.values())[1:]
+            graph[obj_name] = [param.name for param in params]
+
+        return graph
+
+    def _build_classes_from_dependencies(
+        self,
+        dep: DependencyGraph,
+        classes: Dict[str, type],
+        param_map: Dict[str, List[str]]
+    ) -> Dict[str, object]:
+        """
+        Tries to build all classes in the dependency graph. Raises RuntimeError
+        if some dependencies are not available or there was a cyclic dependency.
+        """
+        instances: Dict[str, object] = {}
+        resolved = ChainMap(instances, self.injectables)
+
+        while True:
+            # Find all services with no dependencies (leaves of our graph)
+            leaves = [
+                name for name, dependencies in dep.items() if not dependencies
+            ]
+            if not leaves:
+                if not dep:
+                    return instances
+
+                # Find which dependencies are missing
+                missing = {
+                    d for dependencies in dep.values()
+                    for d in dependencies if d not in dep
+                }
+                if missing:
+                    raise RuntimeError(
+                        f"Some dependencies could not be resolved: {missing}"
+                    )
+                else:
+                    cycle = tuple(dep.keys())
+                    raise RuntimeError(
+                        f"Could not resolve cyclic dependency: {cycle}"
+                    )
+
+            # Build all objects with no dependencies
+            for obj_name in leaves:
+                if obj_name not in resolved:
```

Comment:
I'm not sure I follow

Reply:
Sorry, I meant that this check and one indentation level could probably be lost by moving the `not in resolved` to the construction of `leaves`.
(I said `dep`, but meant `leaves` :man_facepalming: )

Reply:
I see. It doesn't quite work to remove that because on the first round the leaves will contain any injectables. And if some object depends only on an injectable, the injectable has to be "resolved" in the first round which means it will be removed from the unresolved dependencies of other objects. I'd have to duplicate that resolving step for the first round, so I think it's nicer this way where there's no special case for the first round.

In particular `test_dependency_injector_resolves_dependencies` fails with a `cyclic dependency` error when I try to combine them.